Issue with scoped packages

Seems like the package is maintained again, but the issue with scoped packages still exists. 

Obvious culprits are @angular or @ng-bootstrap, e.g. I'm getting this error while trying to install a package: 

```
C:\Program Files\Far Manager>npm set registry http://web-dev-08-zz:5080/

>npm install
npm ERR! code ECONNRESET
npm ERR! errno ECONNRESET
npm ERR! network request to http://web-dev-08-zz:5080/@ng-bootstrap%2fng-bootstrap failed, reason: socket hang up
npm ERR! network This is a problem related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

npm ERR! A complete log of this run can be found in:
npm ERR!  
```

Hi @gabrielcsapo , thanks for your help, my setup is a bit complicated and I don't know npm configuration that well so it's hard to test and verify it's working.
I installed local-npm globally, started it and run "npm set registry http://127.0.0.1:5080"

However I have this in my C:\Users\\[username]\\.npmrc:
```
registry=http://127.0.0.1:5080
@angular:registry=http://registry.npmjs.org/
@ng-bootstrap:registry=http://registry.npmjs.org/
@telerik:registry=http://registry.npmjs.org/
@types:registry=http://registry.npmjs.org/
```

Am I right to assume that only non-scoped packages will be cached by local-npm and scoped ones will be downloaded from the npm registry directly? 
How can I know for sure if the package was cached/loaded from cache or downloaded from npm?
With this setup "npm-install" worked on the machine 1 where local-npm is running.

Hovewer if I update C:\Users\\[username]\\.npmrc to 

```
registry=http://127.0.0.1:5080
@angular:registry=http://127.0.0.1:5080
@ng-bootstrap:registry=http://127.0.0.1:5080
@telerik:registry=http://127.0.0.1:5080
@types:registry=http:http://127.0.0.1:5080
```
And run "npm install" it doesn't work, I get 
```
npm ERR! Cannot read property 'replace' of null
npm ERR! Cannot read property 'replace' of null
npm ERR! Cannot read property 'replace' of null
```
But honestly, I don't know what I'm doing in this case. 

It would be nice to update documentation on npm site to include examples how to work with scoped packages. Also would be good to add some info on how to work with pacakges from different repos, especially if they require authentications, e.g. like @progress: http://www.telerik.com/kendo-angular-ui/getting-started/#toc-add-the-progress-npm-registry 

Also my idea was to use a tool like "local-npm" to have local npm repository that can be used from different machine within networks, build agents etc. I couldn't make it work so far though. 

Cheers

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+## Unreleased
+
+- allows scoped packages to be installed
+
 ## 2.0.0 (05/31/2017)
 
 - fixes issue with request entity too large, by setting the limit of `express-http-proxy` limit to `Infinity`

--- a/lib/index.js
+++ b/lib/index.js
@@ -301,6 +301,108 @@ module.exports = function (argv, callback) {
     });
   });
 
+  // allow support for scoped packages
+  app.get('/tarballs/:user/:package/:version.tgz', function (req, res) {
+      var userName = req.params.user;
+      var pkgName = req.params.package;
+      var pkgVersion = req.params.version;
+      var fullName = `${userName}/${pkgName}`;
+      var id = `${userName}/${pkgName}-${pkgVersion}`;
+
+      logger.time('2: skimLocal.get(' + fullName + ')');
+      getDocument(fullName)
+      .catch(function () {
+        res.status(500).send("you are offline and skimdb isn\'t replicated yet");
+        throw new Error('offline');
+      })
+      .then(function(doc) {
+        logger.timeEnd('2: skimLocal.get(' + fullName + ')');
+        var dist = doc.versions[pkgVersion].dist;
+        logger.time('db.get(' + id + ')');
+        db.get(id, { asBuffer: true, valueEncoding: 'binary' }, function (err, buffer) {
+          logger.timeEnd('db.get(' + id + ')');
+          if (!err) {
+            var hash = crypto.createHash('sha1');
+            hash.update(buffer);
+            if (dist.shasum !== hash.digest('hex')) {
+              // happens when we write garbage to disk somehow
+              logger.warn('hashes don\'t match, not returning');
+            } else {
+              logger.hit(fullName, pkgVersion);
+              return sendBinary(res, buffer);
+            }
+          }
+          logger.miss(fullName, pkgVersion);
+
+          if(dist.info) {
+              logger.time('request.get(' + dist.info + ')');
+              var infoBuffs = [];
+              request.get(dist.info).on('error', function (err) {
+                  logger.info(err);
+                  res.status(500).send('you are offline and this package isn\'t cached');
+              }).pipe(through(function (chunk, _, next) {
+                  infoBuffs.push(chunk);
+                  next();
+              }, function (next) {
+                  next();
+                  var info = JSON.parse(Buffer.concat(infoBuffs));
+
+                  var buffs = [];
+                  request.get(info.dist.tarball).on('error', function (err) {
+                      logger.info(err);
+                      res.status(500).send('you are offline and this package isn\'t cached');
+                  }).pipe(through(function (chunk, _, next) {
+                      buffs.push(chunk);
+                      next();
+                  }, function (next) {
+                      next();
+                      var buffer = Buffer.concat(buffs);
+
+                      logger.timeEnd('request.get(' + dist.info + ')');
+                      sendBinary(res, buffer);
+                      db.put(id, buffer, function (err) {
+                          logger.cached(fullName, pkgVersion);
+                          if (err) {
+                              logger.info(err);
+                          }
+                      });
+                  }));
+              }));
+          } else {
+              logger.time('request.get(' + dist.tarball + ')');
+              var buffs = [];
+              request.get(dist.tarball).on('error', function (err) {
+                logger.info(err);
+                res.status(500).send('you are offline and this package isn\'t cached');
+              }).pipe(through(function (chunk, _, next) {
+                buffs.push(chunk);
+                next();
+              }, function (next) {
+                next();
+                var buffer = Buffer.concat(buffs);
+
+                logger.timeEnd('request.get(' + dist.tarball + ')');
+                sendBinary(res, buffer);
+                db.put(id, buffer, function (err) {
+                  logger.cached(fullName, pkgVersion);
+                  if (err) {
+                    logger.info(err);
+                  }
+                });
+              }));
+          }
+        });
+      })
+      .then(function() {
+        return skimLocal.get(fullName).then(function(doc) {
+          doc.versions[pkgVersion].downloads ? doc.versions[pkgVersion].downloads += 1 : doc.versions[pkgVersion].downloads = 1;
+          return skimLocal.put(doc);
+        });
+      }).catch(function (err) {
+        logger.info(err);
+      });
+  });
+
   app.put('/*', proxy(FAT_REMOTE, {
     limit: Infinity
   }));

--- a/src/package.js
+++ b/src/package.js
@@ -27,7 +27,7 @@ class Package extends React.Component {
               doc: json
           });
       }).catch(function(ex) {
-        console.log('parsing failed', ex)
+        console.log('parsing failed', ex); // eslint-disable-line
       })
   }
   timeFormatter(v, l) {

--- a/test/index.js
+++ b/test/index.js
@@ -10,7 +10,7 @@ const localNpm = require('../lib/index');
 var server = '';
 
 test('local-npm', (t) => {
-    t.plan(14);
+    t.plan(15);
 
     t.test('should write a .npmrc file in the root', (t) => {
       fs.writeFile(path.resolve(__dirname, '..', '.npmrc'), 'registry=http://127.0.0.1:3030/', (err) => {
@@ -163,6 +163,19 @@ test('local-npm', (t) => {
       }).on('error', (e) => {
         t.fail(e);
       });
+    });
+
+    t.test('should be able to fetch a scoped package tarball', (t) => {
+        http.get('http://127.0.0.1:3030/tarballs/@ng-bootstrap/ng-bootstrap/1.0.0-alpha.26.tgz', (res) => {
+          var body = '';
+          res.on('data', (d) => body += d.toString('utf8'));
+          res.on('end', function() {
+            t.equal(body.length, 225450);
+            t.end();
+          });
+        }).on('error', (e) => {
+          t.fail(e);
+        });
     });
 
     t.test('should not be allowed to publish over an already existing version of a module', (t) => {

